server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java

```diff
@@ -0,0 +1,415 @@
+        this.dateHistogramInterval = interval;
+    }
+
+    /** Return the interval as a date time unit if applicable, regardless of how it was configured. If this returns
+     *  {@code null} then it means that the interval is expressed as a fixed
+     *  {@link TimeValue} and may be accessed via {@link #tryIntervalAsFixedUnit()}. */
+    DateTimeUnit tryIntervalAsCalendarUnit() {
+        if (intervalType.equals(IntervalTypeEnum.CALENDAR) || intervalType.equals(IntervalTypeEnum.LEGACY_DATE_HISTO)) {
+            return DateHistogramAggregationBuilder.DATE_FIELD_UNITS.get(dateHistogramInterval.toString());
+        }
+        return null;
+    }
+
+    /**
+     * Get the interval as a {@link TimeValue}, regardless of how it was configured. Returns null if
+     * the interval cannot be parsed as a fixed time.
+     */
+    TimeValue tryIntervalAsFixedUnit() {
+        if (dateHistogramInterval == null || Strings.isNullOrEmpty(dateHistogramInterval.toString())) {
+            return null;
+        }
+        try {
+            return TimeValue.parseTimeValue(dateHistogramInterval.toString(), null, getClass().getSimpleName() + ".interval");
+        } catch (IllegalArgumentException e) {
+            return null;
+        }
+    }
+
+    public Rounding createRounding(ZoneId timeZone) {
+        Rounding.Builder tzRoundingBuilder;
+        if (isEmpty()) {
+            throw new IllegalArgumentException("Invalid interval specified, must be non-null and non-empty");
+        }
+        DateIntervalWrapper.IntervalTypeEnum intervalType = getIntervalType();
+        if (intervalType.equals(DateIntervalWrapper.IntervalTypeEnum.FIXED)) {
+            tzRoundingBuilder = Rounding.builder(tryIntervalAsFixedUnit());
+        } else if (intervalType.equals(DateIntervalWrapper.IntervalTypeEnum.CALENDAR)) {
+            tzRoundingBuilder = Rounding.builder(tryIntervalAsCalendarUnit());
+        } else {
+            // We're not sure what the interval was originally (legacy) so use old behavior of assuming
+            // calendar first, then fixed.  Required because fixed/cal overlap in places ("1h")
+            DateTimeUnit intervalAsUnit = tryIntervalAsCalendarUnit();
+            if (intervalAsUnit != null) {
+                tzRoundingBuilder = Rounding.builder(tryIntervalAsCalendarUnit());
+            } else {
+                tzRoundingBuilder = Rounding.builder(tryIntervalAsFixedUnit());
+            }
+        }
+        if (timeZone != null) {
+            tzRoundingBuilder.timeZone(timeZone);
+        }
+        return tzRoundingBuilder.build();
+    }
+
+    private void setIntervalType(IntervalTypeEnum type) {
+        // If we're the same or have no existing type, just use the provided type
+        if (intervalType.equals(IntervalTypeEnum.NONE) || type.equals(intervalType)) {
+            intervalType = type;
+            return;
+        }
+
+        // interval() method
+        switch (type) {
```

Comment:
> If we're ok breaking DateHistoAggBuilder so that you can only set one of: [interval, dateHistogramInterval, fixedInterval, calendarInterval], it would definitely simplify some of the logic. But I'm not sure we could do that in a minor since it'd break java users (unsure if you can set both in REST)

Ok agreed that we shouldn't break DateHistoAggBuilder here as it would mean a break in a minor. I assume that when we come to removing the legacy interval option this logic will become simple anyway?

Reply:
Correct, this will be simplified considerably once the legacy options are removed.  Then we can just do the mutual exclusion check on the setters themselves and life should be easy.

I'll add some comments laying out the plans so it's clear